cohort-engine/src/main/java/com/ibm/cohort/engine/measure/LibraryHelper.java

```diff
@@ -5,27 +5,171 @@
+			LibraryResolutionProvider<org.hl7.fhir.r4.model.Library> libraryResourceProvider) {
+
+		List<org.cqframework.cql.elm.execution.Library> libraries = new ArrayList<>();
+
+		// TODO: The cqfmeasure IG says there should only be one library here. Should we
+		// enforce that?
+		// http://hl7.org/fhir/us/cqfmeasures/StructureDefinition-measure-cqfm.html
+		for (CanonicalType ref : measure.getLibrary()) {
+			org.hl7.fhir.r4.model.Library library = resolveLibrary(ref.getValue(), libraryLoader,
+					libraryResourceProvider);
+			if (library != null) {
+				libraries.addAll(loadLibraries(library, libraryLoader, libraryResourceProvider));
+			}
+		}
+
+		return libraries;
+	}
+	
+	/**
+	 * Load the provided Library and all of its dependencies using a recursive tree
+	 * walk of related artifacts.
+	 * 
+	 * @param library                 FHIR Library resource
+	 * @param libraryLoader           LibraryLoader library loading mechanism for
+	 *                                handling CQL/ELM source to in memory library
+	 *                                loading
+	 * @param libraryResourceProvider Resource provider that handles loading the
+	 *                                FHIR library resource that contains the
+	 *                                CQL/ELM content
+	 * @return List of executable logic libraries
+	 */
+	public static List<org.cqframework.cql.elm.execution.Library> loadLibraries(org.hl7.fhir.r4.model.Library library,
+			org.opencds.cqf.cql.engine.execution.LibraryLoader libraryLoader,
+			LibraryResolutionProvider<org.hl7.fhir.r4.model.Library> libraryResourceProvider) {
+		Set<String> ids = new HashSet<>();
+		return loadLibraries( library, libraryLoader, libraryResourceProvider, ids );
+	}
+
+	/**
+	 * Load the provided Library and all of its dependencies using a recursive tree
+	 * walk of related artifacts.
+	 * 
+	 * @param library                 FHIR Library resource
+	 * @param libraryLoader           LibraryLoader library loading mechanism for
+	 *                                handling CQL/ELM source to in memory library
+	 *                                loading
+	 * @param libraryResourceProvider Resource provider that handles loading the
+	 *                                FHIR library resource that contains the
+	 *                                CQL/ELM content
+	 * @param loadedIds               Set of library IDs that have already been loaded by the tree walk. This is used for cycle detection.
+	 * @return List of executable logic libraries
+	 */
+	protected static List<org.cqframework.cql.elm.execution.Library> loadLibraries(org.hl7.fhir.r4.model.Library library,
+			org.opencds.cqf.cql.engine.execution.LibraryLoader libraryLoader,
+			LibraryResolutionProvider<org.hl7.fhir.r4.model.Library> libraryResourceProvider, Set<String> loadedIds) {
+
+		List<org.cqframework.cql.elm.execution.Library> libraries = new ArrayList<>();
+		
+		if( ! loadedIds.contains(library.getId()) ) {
```

Comment:
I have a sense that it might be possible for a graph cycle to be non-fatal. Without guard rails in place in the UI, it certainly seems probably that we might end up with cycles and it isn't completely clear to me that they would need to be fatal. Bad design, but maybe not fatal. I'm going to go with the gentle approach for now. Will consider logging a warning.